pkg/test/istioio/snapshot.go

```diff
@@ -0,0 +1,286 @@
+			if err != nil {
+				return fmt.Errorf("failed creating kube client for context %s: %v", contextName, err)
+			}
+
+			clusterSN, err := newClusterSnapshot(client, contextName)
+			if err != nil {
+				return fmt.Errorf("failed creating snapshot for context %s: %v", contextName, err)
+			}
+
+			mutex.Lock()
+			meshSN.Clusters = append(meshSN.Clusters, clusterSN)
+			mutex.Unlock()
+			return nil
+		})
+	}
+
+	if err := wg.Wait(); err != nil {
+		return MeshSnapshot{}, err
+	}
+
+	sort.Slice(meshSN.Clusters, func(i, j int) bool {
+		return strings.Compare(meshSN.Clusters[i].Context, meshSN.Clusters[j].Context) < 0
+	})
+	return meshSN, nil
+}
+
+func newClusterSnapshot(client kube.Client, contextName string) (ClusterSnapshot, error) {
+	mutex := sync.Mutex{}
+	clusterSN := ClusterSnapshot{
+		Context: contextName,
+	}
+	nilVal := ClusterSnapshot{}
+	namespaces, err := client.CoreV1().Namespaces().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nilVal, fmt.Errorf("failed listing namespaces for context %s: %v", contextName, err)
+	}
+
+	var wg errgroup.Group
+	for _, ns := range namespaces.Items {
+		namespace := ns.Name
+
+		// Check whether to include the namespace in the list.
+		if _, ok := namespaceBlacklist[namespace]; !ok {
+			clusterSN.Namespaces = append(clusterSN.Namespaces, namespace)
+		}
+
+		includeKubeResources := kubeResourceWhitelist[namespace]
+		includeIstioResources := istioResourceWhitelist[namespace]
+
+		if !includeKubeResources && !includeIstioResources {
+			// No resources are included for this namespace. Skip it.
+			continue
+		}
+
+		wg.Go(func() error {
+			nsSnapshot := NamespaceSnapshot{
+				Namespace: namespace,
+			}
+
+			if includeKubeResources {
+				// Service
+				if services, err := client.CoreV1().Services(namespace).List(context.TODO(), metav1.ListOptions{}); err != nil {
+					scopes.Framework.Debugf("failed listing services in namespace %s: %v", namespace, err)
+				} else {
+					for _, svc := range services.Items {
+						nsSnapshot.Services = append(nsSnapshot.Services, svc.Name)
+					}
+					sort.Strings(nsSnapshot.Services)
+				}
+
+				// Deployments
+				if deployments, err := client.AppsV1().Deployments(namespace).List(context.TODO(), metav1.ListOptions{}); err != nil {
+					scopes.Framework.Debugf("failed listing deployments in namespace %s: %v", namespace, err)
+				} else {
+					for _, pod := range deployments.Items {
+						nsSnapshot.Deployments = append(nsSnapshot.Deployments, pod.Name)
+					}
+					sort.Strings(nsSnapshot.Deployments)
+				}
+
+				// Pods
+				if pods, err := client.CoreV1().Pods(namespace).List(context.TODO(), metav1.ListOptions{}); err != nil {
+					scopes.Framework.Debugf("failed listing pods in namespace %s: %v", namespace, err)
+				} else {
+					for _, pod := range pods.Items {
+						nsSnapshot.Pods = append(nsSnapshot.Pods, pod.Name)
+					}
+					sort.Strings(nsSnapshot.Pods)
+				}
+			}
+
+			if includeIstioResources {
```

Comment:
One of the problems we had before is that when starting with a clean cluster, the Istio CRDs are not defined, so listing these resources would fail. Does that problem go away with the Go functions that your calling now?

Reply:
No that wouldn't change with this implementation. Just maintaining existing functionality.